web/src/main/java/org/cbioportal/web/CacheController.java

```diff
@@ -45,7 +55,38 @@
         if ("not set".equals(requiredApiKey) || ! requiredApiKey.equals(providedApiKey)) {
             return new ResponseEntity<>("", HttpStatus.UNAUTHORIZED);
         }
-        cacheService.clearCaches(springManagedCache);
+        try {
+            cacheService.clearCaches(springManagedCache);
+        } catch (CacheOperationException e) {
+            return new ResponseEntity<>("Error evicting caches. Please try again or validate correct operation of your configured caching implementation.", HttpStatus.INTERNAL_SERVER_ERROR);
+        }
         return new ResponseEntity<>("Flushed all caches!!!", HttpStatus.OK);
     }
+
+    @RequestMapping(value = "/cache/{studyId}", method = RequestMethod.DELETE, produces = MediaType.TEXT_PLAIN_VALUE)
+    @ApiOperation("Clear and reinitialize caches after import/removal/update of a study")
+    public ResponseEntity<String> clearCachesForStudy(
+        @ApiParam("Secret API key passed in HTTP header. The key is configured in portal.properties of the portal instance.")
+        @RequestHeader(value = "X-API-KEY") String providedApiKey,
+        @PathVariable String studyId,
+        @ApiParam("Clear Spring-managed caches")
+        @RequestParam(defaultValue = "true", required = false) final boolean springManagedCache) {
+        if (!cacheEndpointEnabled) {
+            return new ResponseEntity<>("Cache endpoint is disabled for this instance.", HttpStatus.NOT_FOUND);
+        }
+        if ("not set".equals(requiredApiKey) || ! requiredApiKey.equals(providedApiKey)) {
+            return new ResponseEntity<>("", HttpStatus.UNAUTHORIZED);
+        }
+
+        List<String> allStudyIds = studyRepository.getAllStudies(null, "SUMMARY", null, null, null, null)
+            .stream()
+            .map(study -> study.getCancerStudyIdentifier())
+            .collect(Collectors.toList());
```

Comment:
How do you determine whether a service method receives a piece of information as an argument, or gets it itself by talking to another service? I think I generally choose the second option if I'm confident that the service method in question will always know how to request that information.

For example, in this case, I think `allStudyIds` will always be `studyRepository.getAllStudies`, so I'd do it within the CacheServiceImpl. Interested to hear how you make this sort of determination.

Reply:
Ah good question. I do not have a clear rationale I guess. I like your proposal better I think. I will move the method.